### PR TITLE
Feat: Default Password Policy

### PR DESCRIPTION
**Pull request by podlesrafal**
Feature branch

### CODE REVIEW

File: gravitee-am-management-api/gravitee-am-management-api-rest/src/main/java/io/gravitee/am/management/handlers/management/api/resources/organizations/environments/domains/PasswordPolicyResource.java [hunk 1]
```diff
@@ -107,4 +109,31 @@ public void update(
...
+            @PathParam("domain") String domain,
+            @PathParam("policy") String policy,
+            @Suspended final AsyncResponse response) {
+        final var authenticatedUser = getAuthenticatedUser();
+
+        checkAnyPermission(organizationId, environmentId, domain, Permission.DOMAIN_SETTINGS, Acl.UPDATE)
+                .andThen(domainService.findById(domain)
+                        .switchIfEmpty(Maybe.error(() -> new DomainNotFoundException(domain)))
+                        .flatMapSingle(__ -> passwordPolicyService.setDefaultPasswordPolicy(ReferenceType.DOMAIN, domain, policy, authenticatedUser))
+                        .doOnError(error -> log.error("Update Default Password Policy fails: ", error)))
```
Comment (mstasgravitee) [inline gravitee-am-management-api/gravitee-am-management-api-rest/src/main/java/io/gravitee/am/management/handlers/management/api/resources/organizations/environments/domains/PasswordPolicyResource.java:135]:
```suggestion
                        .doOnError(error -> log.error("Update Default Password Policy fails: {}", error)))
```

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/PasswordPolicyServiceImpl.java [hunk 2]
```diff
@@ -83,20 +81,26 @@ public Flowable<PasswordPolicy> findByDomain(String domain) {
...
+            return Single.error(new InvalidParameterException("Password policy requires a reference type DOMAIN and a reference ID"));
         }
 
         final var now = new Date();
         policy.setCreatedAt(now);
         policy.setUpdatedAt(now);
 
-        return passwordPolicyRepository.create(policy)
+        return passwordPolicyRepository.findByDefaultPolicy(policy.getReferenceType(), policy.getReferenceId())
+                .flatMap(__ -> Maybe.just(policy))
```
Comment (mstasgravitee) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/impl/PasswordPolicyServiceImpl.java:98]:
```suggestion
                .toMaybe()
```
?

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/PasswordPolicyServiceImpl.java [hunk 3]
```diff
@@ -128,23 +132,34 @@ public Single<PasswordPolicy> update(ReferenceType referenceType, String referen
...
+                .switchIfEmpty(Single.error(() -> new PasswordPolicyNotFoundException(policyId)))
+                .flatMap(policyToUpdate -> passwordPolicyRepository.findByDefaultPolicy(referenceType, referenceId)
+                        .flatMapSingle(defaultPolicy -> {
+                            PasswordPolicy nonDefaultPasswordPolicy = new PasswordPolicy(defaultPolicy);
+                            nonDefaultPasswordPolicy.setUpdatedAt(new Date());
+                            nonDefaultPasswordPolicy.setDefaultPolicy(Boolean.FALSE);
+                            return updatePasswordPolicy(referenceType, referenceId, nonDefaultPasswordPolicy, defaultPolicy, principal)
+                                    .flatMap(__ -> setNewDefaultPolicy(referenceType, referenceId, policyToUpdate, principal));
+                        })
+                        .switchIfEmpty(Single.defer(()->setNewDefaultPolicy(referenceType, referenceId, policyToUpdate, principal)))
```
Comment (mstasgravitee) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/impl/PasswordPolicyServiceImpl.java:155]: wrong formatting

File: gravitee-am-ui/src/app/domain/settings/password-policies/domain-password-policies.component.html [hunk 4]
```diff
@@ -42,10 +42,29 @@ <h1>Password Policies</h1>
...
+          <div style="display: flex">
+            <span style="margin-right: 3px">Default</span>
+            <span
+              ><mat-icon
+                style="font-size: 18px; color: black; cursor: pointer"
+                matTooltip="Assign a policy for Identity providers not configured with a specific policy."
+              >
+                info_outline</mat-icon
+              ></span
+            >
```
Comment (mstasgravitee) [inline gravitee-am-ui/src/app/domain/settings/password-policies/domain-password-policies.component.html:56]: this line breaks looks wired

Comment (podlesrafal) [reply gravitee-am-ui/src/app/domain/settings/password-policies/domain-password-policies.component.html:56]: This is linter work :D

File: gravitee-am-management-api/gravitee-am-management-api-rest/src/main/java/io/gravitee/am/management/handlers/management/api/resources/organizations/environments/domains/PasswordPolicyResource.java (hunk 1 above)
Comment (podlesrafal) [reply gravitee-am-management-api/gravitee-am-management-api-rest/src/main/java/io/gravitee/am/management/handlers/management/api/resources/organizations/environments/domains/PasswordPolicyResource.java:135]: In {} we put arguments to string. When we want to log message and throwable we use: `void error(String var1, Throwable var2);`

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/PasswordPolicyServiceImpl.java (hunk 2 above)
Comment (podlesrafal) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/impl/PasswordPolicyServiceImpl.java:98]: Here we dont want to operate on Maybe returned by findByDefaultPolicy. We want to operate on policy from method argument. Thats why we create maybe and then we proceed this in stream. `toMaybe()` will create stream with object from DB.

File: gravitee-am-ui/src/app/domain/settings/password-policies/domain-password-policies.component.html (hunk 4 above)
Comment (podlesrafal) [reply gravitee-am-ui/src/app/domain/settings/password-policies/domain-password-policies.component.html:56]: When I changed it and run `npm run lint:fix` it changed back to this formatting.

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/PasswordPolicyServiceImpl.java (hunk 3 above)
Comment (podlesrafal) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/impl/PasswordPolicyServiceImpl.java]: fixed

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/PasswordPolicyServiceImpl.java (hunk 2 above)
Comment (mstasgravitee) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/impl/PasswordPolicyServiceImpl.java:98]: oh, I can see that Maybe is already return so this line can be remove at all.

File: gravitee-am-ui/src/app/domain/settings/password-policies/domain-password-policies.component.html (hunk 4 above)
Comment (mstasgravitee) [reply gravitee-am-ui/src/app/domain/settings/password-policies/domain-password-policies.component.html:56]: you can put it in one line and then lint. The result will be different

**Review by mstasgravitee — APPROVED** [review]

**Pull request merged**